autoretrieve/engine.go

```diff
@@ -0,0 +1,524 @@
+		log.Info("Skipped announcing the latest: no previously published advertisements.")
+		return nil
+	}
+	log.Infow("Republishing latest advertisement", "cid", adCid)
+
+	return e.publisher.UpdateRoot(ctx, adCid)
+}
+
+// RegisterMultihashLister registers a provider.MultihashLister that is used to look up the
+// list of multihashes associated to a context ID. At least one such registration
+// must be registered before calls to AutoretrieveEngine.NotifyPut and AutoretrieveEngine.NotifyRemove.
+//
+// Note that successive calls to this function will replace the previous registration.
+// Only a single registration is supported.
+//
+// See: provider.Interface
+func (e *AutoretrieveEngine) RegisterMultihashLister(mhl provider.MultihashLister) {
+	log.Debugf("Registering multihash lister in engine")
+	e.cblk.Lock()
+	defer e.cblk.Unlock()
+	e.mhLister = mhl
+}
+
+// NotifyPut publishes an advertisement that signals the list of multihashes
+// associated to the given contextID is available by this provider with the
+// given metadata. A provider.MultihashLister is required, and is used to look up the
+// list of multihashes associated to a context ID.
+//
+// Note that prior to calling this function a provider.MultihashLister must be registered.
+//
+// See: AutoretrieveEngine.RegisterMultihashLister, AutoretrieveEngine.Publish.
+func (e *AutoretrieveEngine) NotifyPut(ctx context.Context, contextID []byte, providerID string, addresses []string, md metadata.Metadata) (cid.Cid, error) {
+	// The multihash lister must have been registered for the linkSystem to know how to
+	// go from contextID to list of CIDs.
+	return e.publishAdvForIndex(ctx, contextID, providerID, addresses, md, false)
+}
+
+// NotifyRemove publishes an advertisement that signals the list of multihashes associated to the given
+// contextID is no longer available by this provider.
+//
+// Note that prior to calling this function a provider.MultihashLister must be registered.
+//
+// See: AutoretrieveEngine.RegisterMultihashLister, AutoretrieveEngine.Publish.
+func (e *AutoretrieveEngine) NotifyRemove(ctx context.Context, contextID []byte) (cid.Cid, error) {
+	return e.publishAdvForIndex(ctx, contextID, "", nil, metadata.Metadata{}, true)
+}
+
+// Shutdown shuts down the engine and discards all resources opened by the engine.
+// The engine is no longer usable after the call to this function.
+func (e *AutoretrieveEngine) Shutdown() error {
+	var errs error
+	if e.publisher != nil {
+		if err := e.publisher.Close(); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("error closing leg publisher: %s", err))
+		}
+	}
+	if err := e.entriesChunker.Close(); err != nil {
+		errs = multierror.Append(errs, fmt.Errorf("error closing link entriesChunker: %s", err))
+	}
+	return errs
+}
+
+// GetAdv gets the advertisement associated to the given cid c.
+// The context is not used.
+func (e *AutoretrieveEngine) GetAdv(_ context.Context, adCid cid.Cid) (*schema.Advertisement, error) {
```

Comment:
if context.Context is not used, maybe remove it

Reply:
Given that interface was created in the reference repo, as the engine is a different package from the provider. This implementation is not the case, we should be fine without it